### PR TITLE
Fix: lwc compiler errors on .jsx/.tsx files

### PR DESCRIPTION
**Pull request by SaurabhMulay999**
## Details
Fix: LWC compiler errors on .jsx/.tsx files #4122. 
(Treating .jsx/.tsx  file extensions the same as .js & .ts.) 

## Does this pull request introduce a breaking change?
-   😮‍💨 No, it does not introduce a breaking change.

## Does this pull request introduce an observable change?
-   🤞 No, it does not introduce an observable change.

### CODE REVIEW

**Comment by nolanlawson** [comment]
Hi @SaurabhMulay999 thank you for the contribution! Could you please add a test to confirm the fix? The right place to add a test would be in [`transform-javascript.spec.ts`](https://github.com/salesforce/lwc/blob/32f4c5728a13c561c798d520c68ba3a0877043b7/packages/%40lwc/compiler/src/transformers/__tests__/transform-javascript.spec.ts). We just need to confirm that the new file extensions are handled correctly.

**Comment by SaurabhMulay999** [comment]
> Hi @SaurabhMulay999 thank you for the contribution! Could you please add a test to confirm the fix? The right place to add a test would be in [`transform-javascript.spec.ts`](https://github.com/salesforce/lwc/blob/32f4c5728a13c561c798d520c68ba3a0877043b7/packages/%40lwc/compiler/src/transformers/__tests__/transform-javascript.spec.ts). We just need to confirm that the new file extensions are handled correctly.

Sure, I'll try.

**Comment by SaurabhMulay999** [comment]
Hi @nolanlawson ,
I have tried writing a unit test. 

Code:

describe('transformFile function',()=>{
    it('should be tranform Typescript React file (.tsx)',()=>{
        const src=`import React from 'react'; \n const App=()=><div>Hello World</div>;`
        const fileName='app.tsx';
        const options={namespace:'c',name:'app'};
        const result=transformSync(src,fileName,options);
        expect(result.code).toBeDefined();
        expect(result.map).toBeDefined();
        expect(result.warnings).toBeUndefined();
        expect(result.cssScopeTokens).toBeUndefined();

    });

    it('should be tranform Typescript React file (.jsx)',()=>{
        const src=`import React from 'react'; \n const App=()=><div>Hello World</div>;`
        const fileName='app.jsx';

        const options={namespace:'c',name:'app'};
        const result=transformSync(src,fileName,options);
        expect(result.code).toBeDefined();
        expect(result.map).toBeDefined();
        expect(result.warnings).toBeUndefined();
        expect(result.cssScopeTokens).toBeUndefined();

    });
});

after debug it was failing saying that. 

Error:
Support for the experimental syntax 'jsx' isn't currently enabled.qq

I have checked some online resources it was suggesting to install @babel/preset-react.

**Comment by nolanlawson** [comment]
@SaurabhMulay999 It should be fine to just put actual JavaScript in there rather than JSX.

We do not need to test transpiling JSX into JavaScript; just that the `.jsx`/`.tsx` extensions are supported.

**Comment by SaurabhMulay999** [comment]
> @SaurabhMulay999 It should be fine to just put actual JavaScript in there rather than JSX.
> 
> We do not need to test transpiling JSX into JavaScript; just that the `.jsx`/`.tsx` extensions are supported.
> 
> In a real app setup (e.g. using Rollup), the consumer of LWC would be responsible for ensuring that there are plugins in place to do the transformation.

Okay. I've added the JS format. I have ran it locally it was successful. Let me know if any modifications are required.

File: packages/@lwc/compiler/src/transformers/__tests__/transform-javascript.spec.ts [hunk 1]
```diff
@@ -269,3 +269,34 @@ describe('sourcemaps', () => {
         });
     });
 });
+
+describe('transformFile function', () => {
+    it('should be tranform Typescript file with ext (.tsx)', () => {
+        const src = `
+        import { LightningElement } from 'lwc';
+        export default class Foo extends LightningElement {}
+    `;
```
Comment (nolanlawson) [inline packages/@lwc/compiler/src/transformers/__tests__/transform-javascript.spec.ts:278]:
```suggestion
        const src = `
            import { LightningElement } from 'lwc';
            export default class Foo extends LightningElement {}
        `;
```

Just slightly nicer formatting.

File: packages/@lwc/compiler/src/transformers/__tests__/transform-javascript.spec.ts [hunk 2]
```diff
@@ -269,3 +269,34 @@ describe('sourcemaps', () => {
...
+        expect(result.map).toBeDefined();
+        expect(result.warnings).toBeUndefined();
+        expect(result.cssScopeTokens).toBeUndefined();
+    });
+
+    it('should be tranform Javascript file with ext (.jsx)', () => {
+        const src = `
+        import { LightningElement } from 'lwc';
+        export default class Foo extends LightningElement {}
+        `;
```
Comment (nolanlawson) [inline packages/@lwc/compiler/src/transformers/__tests__/transform-javascript.spec.ts:292]:
```suggestion
        const src = `
            import { LightningElement } from 'lwc';
            export default class Foo extends LightningElement {}
        `;
```

Ditto.

File: packages/@lwc/compiler/src/transformers/__tests__/transform-javascript.spec.ts [hunk 3]
```diff
@@ -269,3 +269,34 @@ describe('sourcemaps', () => {
...
     });
 });
+
+describe('transformFile function', () => {
+    it('should be tranform Typescript file with ext (.tsx)', () => {
+        const src = `
+        import { LightningElement } from 'lwc';
+        export default class Foo extends LightningElement {}
+    `;
+        const fileName = 'app.ts';
```
Comment (nolanlawson) [inline packages/@lwc/compiler/src/transformers/__tests__/transform-javascript.spec.ts:279]:
```suggestion
        const fileName = 'app.tsx';
```

File: packages/@lwc/compiler/src/transformers/__tests__/transform-javascript.spec.ts [hunk 4]
```diff
@@ -269,3 +269,34 @@ describe('sourcemaps', () => {
...
+        expect(result.warnings).toBeUndefined();
+        expect(result.cssScopeTokens).toBeUndefined();
+    });
+
+    it('should be tranform Javascript file with ext (.jsx)', () => {
+        const src = `
+        import { LightningElement } from 'lwc';
+        export default class Foo extends LightningElement {}
+        `;
+        const fileName = 'app.js';
```
Comment (nolanlawson) [inline packages/@lwc/compiler/src/transformers/__tests__/transform-javascript.spec.ts:293]:
```suggestion
        const fileName = 'app.jsx';
```

**Comment by SaurabhMulay999** [comment]
> @SaurabhMulay999 Could you please sign the CLA as described above?

Hey @nolanlawson,
I don't know what is the issue but previously i already have signed the cla. that site redirect me and shows that, that i already have signed the cla. but this time system failed to verify my email. is there any workaround? like to resign it.

**Comment by wjhsf** [comment]
> Commit authors must be associated with GitHub users

I think you'll have to redo the commits using the email address associated with your GitHub account. (Or add the email address from your commits to your GitHub account, but the one used [here](https://github.com/salesforce/lwc/commit/c222bd30c398359ede528aa41d25a0d872aa70a1.patch) doesn't seem to be a real email address, so I don't think GitHub will let you.)

**Review by ekashida — COMMENTED** [review]
It would be nice to loop over an array of all the supported file extensions. Something like:
```js

    describe('file extension support', () => {
        function testFileExtensionSupport(ext: string) {
            it(`should support ${ext} file extension`, () => {
                const src = `
                    import { LightningElement } from 'lwc';
                    export default class Foo extends LightningElement {}
                `;
                const options = { namespace: 'c', name: 'foo' };
                const result = transformSync(src, `foo${ext}`, options);
                expect(result.code).toBeDefined();
                expect(result.map).toBeDefined();
                expect(result.warnings).toBeUndefined();
                expect(result.cssScopeTokens).toBeUndefined();
            });
        }
        ['.js', '.jsx', '.ts', '.tsx'].forEach(testFileExtensionSupport);
    });
```

**Comment by SaurabhMulay999** [comment]
> > Commit authors must be associated with GitHub users
> 
> I think you'll have to redo the commits using the email address associated with your GitHub account. (Or add the email address from your commits to your GitHub account, but the one used [here](https://github.com/salesforce/lwc/commit/c222bd30c398359ede528aa41d25a0d872aa70a1.patch) doesn't seem to be a real email address, so I don't think GitHub will let you.)

I think it tried to fetch my official salesforce email. I'll try adding the personal email and then recommit.

File: packages/@lwc/compiler/src/transformers/__tests__/transform-javascript.spec.ts [hunk 5]
```diff
@@ -269,3 +269,34 @@ describe('sourcemaps', () => {
         });
     });
 });
+
+describe('transformFile function', () => {
```
[the remark this answers is not in the document]
Comment (SaurabhMulay999) [reply packages/@lwc/compiler/src/transformers/__tests__/transform-javascript.spec.ts:273]: I think it's better as you have suggested above to loop over all the extensions.

**Comment by wjhsf** [comment]
> I think it tried to fetch my official salesforce email. I'll try adding the personal email and then recommit.
If you use your salesforce email then you don't need to sign the CLA, because it will recognize you as an internal user.

**Comment by SaurabhMulay999** [comment]
> > I think it tried to fetch my official salesforce email. I'll try adding the personal email and then recommit.
> > If you use your salesforce email then you don't need to sign the CLA, because it will recognize you as an internal user.

Yeah, but this is my personal account. I have added the Name as well email in the latest commit. It should have now recognise me as a GitHub user.

**Review by ekashida — APPROVED** [review]
lgtm!

**Comment by SaurabhMulay999** [comment]
> > Yeah, but this is my personal account. I have added the Name as well email in the latest commit. It should have now recognise me as a GitHub user.
> 
> The older commits still show the invalid email address. I believe you will need to do an [interactive rebase](https://www.sitepoint.com/git-interactive-rebase-guide/) or squash-merge followed by a force-push to fix this.

Okay. I'll try interactive rebase.

**Comment by SaurabhMulay999** [comment]
> > Yeah, but this is my personal account. I have added the Name as well email in the latest commit. It should have now recognise me as a GitHub user.
> 
> The older commits still show the invalid email address. I believe you will need to do an [interactive rebase](https://www.sitepoint.com/git-interactive-rebase-guide/) or squash-merge followed by a force-push to fix this.

Hi @nolanlawson , 
I have changed the commit author for all the commits. and also have squashed (2 to N) commits together. Thanks for the suggestion, I have learned a lot.

File: packages/@lwc/compiler/src/transformers/__tests__/transform-javascript.spec.ts [hunk 6]
```diff
@@ -269,3 +269,21 @@ describe('sourcemaps', () => {
...
+            `;
+            const options = { namespace: 'c', name: 'foo' };
+            const result = transformSync(src, `foo${ext}`, options);
+            expect(result.code).toBeDefined();
+            expect(result.map).toBeDefined();
+            expect(result.warnings).toBeUndefined();
+            expect(result.cssScopeTokens).toBeUndefined();
+        });
+    }
+    ['.js', '.jsx', '.ts', '.tsx'].forEach(testFileExtensionSupport);
```
Comment (nolanlawson) [inline packages/@lwc/compiler/src/transformers/__tests__/transform-javascript.spec.ts:288]: Nice!

**Review by nolanlawson — APPROVED** [review]

**Pull request merged**